bkbit/scripts/gfftranslator.py

```diff
@@ -1,203 +1,682 @@
+            curr_line_num = 1
+            progress_bar = tqdm(
+                total=self.__get_line_count(gff_file), desc="Parsing GFF3 File"
+            )
+            for line_raw in file:
+                line_strip = line_raw.strip()
+                if curr_line_num == 1 and not line_strip.startswith("##gff-version 3"):
+                    logger.critical(
+                        'Line %s: ##gff-version 3" missing from the first line.',
+                        curr_line_num,
+                    )
+                elif len(line_strip) == 0:  # blank line
+                    continue
+                elif line_strip.startswith("##"):  # TODO: parse more metadata
+                    pass
+                elif line_strip.startswith("#"):  # TODO: parse more metadata
+                    pass
+                else:  # line may be a feature or unknown
+                    tokens = list(map(str.strip, line_raw.split("\t")))
+                    if len(tokens) != 9:
+                        logger.warning(
+                            "Line %s: Features are expected 9 columns, found %s.",
+                            curr_line_num,
+                            len(tokens),
+                        )
+                    if (
+                        tokens[2] in feature_filter
+                    ):  # only look at rows that have a type that is included in feature_filter
+                        attributes = self.__merge_values(
+                            tuple(a.split("=") for a in tokens[8].split(";"))
+                        )
+                        # TODO: Write cleaner code that calls respective generate function based on the authority automatically
+                        if (
+                            self.genome_annotation.authority
+                            == kbmodel.AuthorityType.ENSEMBL
+                        ):
+                            gene_annotation = self.generate_ensembl_gene_annotation(
+                                attributes, curr_line_num
+                            )
+                            if gene_annotation is not None:
+                                self.gene_annotations[gene_annotation] = gene_annotation
+                        elif (
+                            self.genome_annotation.authority
+                            == kbmodel.AuthorityType.NCBI
+                        ):
+                            gene_annotation = self.generate_ncbi_gene_annotation(
+                                attributes, curr_line_num
+                            )
+                            if gene_annotation is not None:
+                                self.gene_annotations[gene_annotation.id] = (
+                                    gene_annotation
+                                )
+                progress_bar.update(1)
+                curr_line_num += 1
+            progress_bar.close()
+
+    def generate_ensembl_gene_annotation(self, attributes, curr_line_num):
```

Comment:
I'm not asking you to change this, but:

If `generate_ensembl_gene_annotation` and `generate_ncbi_gene_annotation` had been functions independent of the `Gff3` class that just accept `attributes` dicts and return `GeneAnnotation` objects, they would have been easier to test (i.e. you would have been able to test *just* their functionality without having to write out an entire well-formed GFF3 file).

Again: I'm not asking for this change now, but if we ever get into a situation where we think breaking these methods out will make our lives simpler, I think we should be open to that.

Reply:
I agree, I will keep note of this.